### PR TITLE
TaskService: Add mysql based task storage

### PR DESCRIPTION
**Pull request by w-zr**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #5240 

## What this PR does / why we need it:

Add mysql based task storage.

### CODE REVIEW

File: pkg/taskservice/mysql_task_storage.go [hunk 1]
```diff
@@ -0,0 +1,580 @@
...
+    						task_retry_interval=?,
+    						task_delay_duration=?,
+    						cron_expr=?,
+    						next_time=?,
+    						trigger_times=?,
+    						create_at=?,
+    						update_at=? where cron_task_id=?`
+)
+
+// used for testing
```
Comment (zhangxu19830126) [inline pkg/taskservice/mysql_task_storage.go:154]: remove comments

File: pkg/taskservice/mysql_task_storage.go [hunk 2]
```diff
@@ -0,0 +1,580 @@
...
+type mysqlTaskStorage struct {
+	db *sql.DB
+
+	// Used for testing. Make some changes to the data before updating.
```
Comment (zhangxu19830126) [inline pkg/taskservice/mysql_task_storage.go:158]: it's not needed in mysql based task storage

File: pkg/taskservice/mysql_task_storage.go [hunk 3]
```diff
@@ -0,0 +1,580 @@
...
+func NewMysqlTaskStorage(driver, dsn string) TaskStorage {
+	db, err := sql.Open(driver, dsn)
+	if err != nil {
+		panic(err)
+	}
+
+	for _, s := range createSqls {
+		_, err = db.Exec(s)
+		if err != nil {
+			panic(err)
```
Comment (zhangxu19830126) [inline pkg/taskservice/mysql_task_storage.go:172]: Add retry, can not panic directly

File: pkg/taskservice/mysql_task_storage.go [hunk 4]
```diff
@@ -0,0 +1,580 @@
...
+		db: db,
+
+		preUpdateCron: func() error { return nil },
+		preUpdate:     func() {},
+	}
+}
+
+func (m *mysqlTaskStorage) Close() error {
+	err := m.db.Close()
+	if err != nil {
```
Comment (zhangxu19830126) [inline pkg/taskservice/mysql_task_storage.go:186]: handle error like this:  if err := xxx(); err != nil {
    return err
}

File: pkg/taskservice/mysql_task_storage.go [hunk 5]
```diff
@@ -0,0 +1,580 @@
...
+		return err
+	}
+	return nil
+}
+
+func (m *mysqlTaskStorage) Add(ctx context.Context, tasks ...task.Task) (int, error) {
+	n := 0
+	for _, t := range tasks {
```
Comment (zhangxu19830126) [inline pkg/taskservice/mysql_task_storage.go:194]: all tasks must in a tranasction.

File: pkg/taskservice/mysql_task_storage.go [hunk 6]
```diff
@@ -0,0 +1,580 @@
...
+
+var (
+	createSqls = []string{
+		`create table if not exists sys_async_task (
+			task_id                     int primary key auto_increment,
+			task_metadata_id            varchar(16) unique not null,
+			task_metadata_executor      int,
+			task_metadata_context       blob,
+			task_max_retry_times        int,
+			task_retry_interval         int,
```
Comment (zhangxu19830126) [inline pkg/taskservice/mysql_task_storage.go:35]: task options and task exec result only one field is needed to store the json string inside

File: pkg/taskservice/mysql_task_storage.go [hunk 7]
```diff
@@ -0,0 +1,580 @@
...
+	return nil
+}
+
+func (m *mysqlTaskStorage) Add(ctx context.Context, tasks ...task.Task) (int, error) {
+	n := 0
+	for _, t := range tasks {
+		if m.taskExists(t.Metadata.ID) {
+			continue
+		}
+		exec, err := m.db.Exec(
```
Comment (zhangxu19830126) [inline pkg/taskservice/mysql_task_storage.go:198]: Need to handle index constraint violation errors, metadata id is unique

File: pkg/taskservice/mysql_task_storage.go [hunk 8]
```diff
@@ -0,0 +1,580 @@
...
+		return 0, nil
+	}
+	triggerTimes, err := m.getTriggerTimes(cronTask.Metadata.ID)
+	if err == sql.ErrNoRows || triggerTimes != cronTask.TriggerTimes-1 {
+		return 0, nil
+	}
+
+	update, err := m.Add(ctx, t)
+	if err != nil {
+		return 0, err
```
Comment (zhangxu19830126) [inline pkg/taskservice/mysql_task_storage.go:470]: update and add need in a transaction

File: pkg/taskservice/mysql_task_storage.go [hunk 9]
```diff
@@ -0,0 +1,580 @@
...
+	}
+	affected, err := exec.RowsAffected()
+	if err != nil {
+		panic(err)
+	}
+
+	return update + int(affected), nil
+}
+
+func (m *mysqlTaskStorage) drop(ctx context.Context) error {
```
Comment (zhangxu19830126) [inline pkg/taskservice/mysql_task_storage.go:495]: why we need drop?

**Review by zhangxu19830126 — CHANGES_REQUESTED** [review]

File: pkg/taskservice/mysql_task_storage.go (hunk 2 above)
Comment (w-zr) [reply pkg/taskservice/mysql_task_storage.go:158]: removed.

File: pkg/taskservice/mysql_task_storage.go (hunk 4 above)
Comment (w-zr) [reply pkg/taskservice/mysql_task_storage.go:186]: fixed.

File: pkg/taskservice/mysql_task_storage.go (hunk 1 above)
Comment (w-zr) [reply pkg/taskservice/mysql_task_storage.go:154]: fixed

File: pkg/taskservice/mysql_task_storage.go (hunk 3 above)
Comment (w-zr) [reply pkg/taskservice/mysql_task_storage.go:172]: return err directly.

File: pkg/taskservice/mysql_task_storage.go (hunk 6 above)
Comment (w-zr) [reply pkg/taskservice/mysql_task_storage.go:35]: fixed.

File: pkg/taskservice/mysql_task_storage.go (hunk 9 above)
Comment (w-zr) [reply pkg/taskservice/mysql_task_storage.go:495]: fixed.

**Review by zhangxu19830126 — APPROVED** [review]

File: pkg/taskservice/mysql_task_storage.go [hunk 10]
```diff
@@ -0,0 +1,640 @@
...
+	}
+	return ""
+}
+
+func removeDuplicateTasks(err error, tasks []task.Task) ([]task.Task, error) {
+	me, ok := err.(*mysql.MySQLError)
+	if !ok {
+		return nil, err
+	}
+	if me.Number != 1062 {
```
Comment (fengttt) [inline pkg/taskservice/mysql_task_storage.go:607]: What is 1062?

File: pkg/taskservice/mysql_task_storage.go [hunk 11]
```diff
@@ -0,0 +1,640 @@
...
+	"github.com/go-sql-driver/mysql"
+
+	"github.com/matrixorigin/matrixone/pkg/pb/task"
+)
+
+var (
+	createSqls = []string{
+		`create table if not exists sys_async_task (
+			task_id                     int primary key auto_increment,
+			task_metadata_id            varchar(16) unique not null,
```
Comment (fengttt) [inline pkg/taskservice/mysql_task_storage.go:33]: Where did the 16 and 20 come from?   I see 1000 is a "big" number that is enough for your use.

But varchar(16)?   bigint(20)?   What at those, esp bigint -- why not just bigint?

File: pkg/taskservice/mysql_task_storage.go (hunk 10 above)
Comment (w-zr) [reply pkg/taskservice/mysql_task_storage.go:607]: It's mysql's error code. Replace it with moerr.ER_DUP_ENTRY.

File: pkg/taskservice/mysql_task_storage.go (hunk 11 above)
Comment (w-zr) [reply pkg/taskservice/mysql_task_storage.go:33]: varchar(16) is the length of a uuid.
20 is an arbitrary number. just bigint is ok.

**Pull request merged**